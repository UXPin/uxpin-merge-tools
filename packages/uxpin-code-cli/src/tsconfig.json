{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "sourceMap": true,
    "strict": true,
    "moduleResolution": "node",
    "baseUrl": "../",
    "skipLibCheck": true,
    "outDir": "../dist/",
    "paths": {
      "*": [
        "node_modules/*",
        "src/types/*"
      ]
    }
  }
}