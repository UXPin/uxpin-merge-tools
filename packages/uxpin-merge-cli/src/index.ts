import 'source-map-support/register';
export { runProgram } from './program/runProgram';
export { Command } from './program/command/Command';
export { getToolVersion } from './program/utils/version/getToolVersion';
export { buildDesignSystemWithStorybook } from './steps/building/buildDesignSystemWithStorybook';
