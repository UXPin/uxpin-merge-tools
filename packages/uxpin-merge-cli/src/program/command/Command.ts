export enum Command {
  DUMP = 'dump',
  EXPERIMENT = 'experiment',
  INIT = 'init',
  GENERATE_PRESETS = 'generate-presets',
  PUSH = 'push',
  DELETE_VERSION = 'delete-version',
  SERVER = 'server',
  SUMMARY = 'summary',
}

export const DEFAULT_COMMAND: Command = Command.EXPERIMENT;
