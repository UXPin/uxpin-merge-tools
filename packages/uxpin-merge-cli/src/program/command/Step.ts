import { DSMetadata } from '../DSMeta';

export interface Step {
  shouldRun: boolean;
  exec: (infos: DSMetadata) => any;
}

export type StepExecutor = (designSystem: DSMetadata) => Promise<DSMetadata>;

export type StepWithoutDSExecutor = () => Promise<void>;
