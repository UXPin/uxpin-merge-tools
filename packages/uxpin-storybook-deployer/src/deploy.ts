import { PresignedPost } from 'aws-sdk/clients/s3';
import axios, { AxiosError, AxiosPromise, AxiosRequestConfig, AxiosResponse } from "axios";
import { execSync } from 'child_process';
import { createReadStream, readdirSync, statSync } from 'fs';
import { relative, join } from 'path';
import FormData from 'form-data';
import { IncomingMessage } from 'http';
import chalk from 'chalk';
import mime from 'mime-types';

type DeployOptions = {
  source: string,
  token: string,
  uxpinDomain?: string,
}

type GetLibraryHashByTokenResponse = {
  libraryHash: string,
}

type UpdateCommitHashBody = {
  commitHash: string,
}

const UXPIN_DOMAIN = 'uxpin.com';
const AUTH_TOKEN_HEADER_NAME = 'auth-token';

let latestCommitHash:string;

export async function deploy(args: DeployOptions):Promise<void> {
  try {
    latestCommitHash = getLatestCommitHash(args.source);
    const { libraryHash } = await getLibraryHashByToken(args);
    const presignedS3PostUrlResponse = await getPresignedS3PostUrl(args);
    await deployStorybookToS3(args, libraryHash, presignedS3PostUrlResponse);
    console.log('Updating commit hash on UXPin...');
    await updateCommitHash(args);
    const storybookUrl = join(getBaseUrl(args.uxpinDomain), 'libraries-storybook', libraryHash, 'index.html');
    console.log(chalk.green(`âœ… DONE! Your storybook is available on ${storybookUrl}`));
  } catch (error) {
    typeof(error) === 'string' ? console.log(chalk.red(error)) : console.log(error);
  }
}

function getBaseUrl(uxpinDomain:string | undefined):string {
  return `https://api.${uxpinDomain || UXPIN_DOMAIN}`;
}

async function getLibraryHashByToken(args:DeployOptions):Promise<GetLibraryHashByTokenResponse> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/getLibraryHashByToken`;
  return axiosWithEnhancedError({ url, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token, method: 'GET' }})
    .then((res:AxiosResponse) => res.data as GetLibraryHashByTokenResponse);
}

async function getPresignedS3PostUrl(args:DeployOptions):Promise<PresignedPost> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/getPresignedS3PostUrl/${latestCommitHash}`;
  return axiosWithEnhancedError({ url, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token, method: 'GET' }})
    .then((res:AxiosResponse) => res.data as PresignedPost);
}

function getLatestCommitHash(source:string):string {
  try {
    return execSync(`cd ${source} && git rev-parse HEAD`).toString().trim();
  } catch (error) {
    throw `ERROR: Failed to get commit hash from ${source}. Please make sure you can run\ngit rev-parse HEAD \ncommand in the directory because UXPin requires latest commit hash`;
  }
}

function getS3Prefix(libraryHash:string):string {
  return `${libraryHash}/${latestCommitHash}/`;
}

function getFilesList(source:string):string[] {
  let results:string[] = [];
  const list:string[] = readdirSync(source);
  list.forEach((fileOrDir) => {
      fileOrDir = source + '/' + fileOrDir;
      const stat = statSync(fileOrDir);
      if (stat && stat.isDirectory()) {
        results = results.concat(getFilesList(fileOrDir));
      } else {
        results.push(fileOrDir);
      }
    });
  return results;
}

async function deployStorybookToS3(args:DeployOptions, libraryHash:string, presignedS3urlResponse:PresignedPost):Promise<void[]> {
  console.log('Start uploading Storybook assets...');
  const s3Prefix = getS3Prefix(libraryHash);
  const promises:Array<Promise<void>> = getFilesList(args.source).map(async (absFilePath) => {
    const relativePath = relative(args.source, absFilePath);
    const key = join(s3Prefix, relativePath);

    const form = new FormData();
    Object.entries(presignedS3urlResponse.fields).forEach(([field, value]) => {
      form.append(field, value);
    });
    form.append('key', key);
    form.append('Content-Type', mime.contentType(mime.lookup(absFilePath) || 'application/octet-stream'));
    form.append('file', createReadStream(absFilePath));

    return new Promise((resolve, reject) => form.submit(presignedS3urlResponse.url, (err, data:IncomingMessage) => {
      if (err) {
        reject(err);
      }
      else if (data.statusCode !== 204) {
        reject({
          status: data.statusCode,
          statusMessage: data.statusMessage,
          filePath: relativePath,
        });
      }
      resolve();
    }));
  });

  return Promise.all(promises);
}

async function updateCommitHash(args:DeployOptions):Promise<boolean> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/updateCommitHash`;
  const data:UpdateCommitHashBody = { commitHash: latestCommitHash };
  return axiosWithEnhancedError({ url, data, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token }, method: 'PUT'})
    .then(() => true);
}

function axiosWithEnhancedError(options:AxiosRequestConfig):AxiosPromise {
  return axios(options).catch((error:AxiosError) => {
    if (error.response) {
      throw {
        ...error.response.data,
        url: options.url,
      };
    }
    throw error;
  });
}
