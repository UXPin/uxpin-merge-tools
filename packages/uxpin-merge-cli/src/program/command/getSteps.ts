import { ProgramArgs } from '../args/ProgramArgs';
import { Command } from './Command';
import { getDeleteVersionCommandSteps } from './delete_version/getDeleteVersionCommandSteps';
import { getDumpCommandSteps } from './dump/getDumpCommandSteps';
import { getExperimentationCommandSteps } from './experimentation/getExperimentationCommandSteps';
import { getGeneratePresetsCommandSteps } from './generate_presets/getGeneratePresetsCommandSteps';
import { getInitCommandSteps } from './init/getInitCommandSteps';
import { getPushCommandSteps } from './push/getPushCommandSteps';
import { getServerCommandSteps } from './server/getServerCommandSteps';
import { Step } from './Step';
import { getSummaryCommandSteps } from './summary/getSummaryCommandSteps';

export function getSteps(args: ProgramArgs): Step[] {
  switch (args.command) {
    case Command.DUMP:
      return getDumpCommandSteps();

    case Command.EXPERIMENT:
      return getExperimentationCommandSteps(args);

    case Command.INIT:
      return getInitCommandSteps(args);

    case Command.GENERATE_PRESETS:
      return getGeneratePresetsCommandSteps(args);

    case Command.PUSH:
      return getPushCommandSteps(args);

    case Command.DELETE_VERSION:
      return getDeleteVersionCommandSteps(args);

    case Command.SERVER:
      return getServerCommandSteps(args);

    case Command.SUMMARY:
      return getSummaryCommandSteps();

    default:
      return [];
  }
}
