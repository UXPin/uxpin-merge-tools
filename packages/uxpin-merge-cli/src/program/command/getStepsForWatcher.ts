import { ProgramArgs } from '../args/ProgramArgs';
import { Command } from './Command';
import { getExperimentationWatchCommandSteps } from './experimentation/getExperimentationWatchCommandSteps';
import { Step } from './Step';

export function getStepsForWatcher(args: ProgramArgs): Step[] {
  if (args.command === Command.EXPERIMENT) {
    return getExperimentationWatchCommandSteps(args);
  }

  return [];
}
