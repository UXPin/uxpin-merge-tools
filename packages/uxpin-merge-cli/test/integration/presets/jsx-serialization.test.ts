import { Command } from '../../../src';
import { runUXPinMergeCommand } from '../../utils/command/runUXPinMergeCommand';
import { setTimeoutBeforeAll } from '../../utils/command/setTimeoutBeforeAll';

const CURRENT_TIMEOUT = 60000;
setTimeoutBeforeAll(CURRENT_TIMEOUT);

jest.mock('../../../src/program/utils/version/getToolVersion');

describe('The dump command', () => {
  describe('run for the `withPresets` design system', () => {
    it('prints the JSON including serialized presets', async () => {
      // given
      // when
      const output: string = await runUXPinMergeCommand({
        cwd: 'resources/designSystems/withPresets',
        params: [Command.DUMP, '--config "uxpin.config.js"', '--webpack-config "./webpack.config.js"'],
      });

      // then
      expect(JSON.parse(output)).toMatchSnapshot({
        vcs: expect.objectContaining({
          branchName: expect.any(String),
          commitHash: expect.stringMatching(/[a-z0-9]+/),
        }),
      });
    });
  });

  describe('run for the `withCorruptedPresets` design system', () => {
    it('prints error about unresolved path', async () => {
      expect.assertions(1);

      // given
      // when
      try {
        await runUXPinMergeCommand({
          cwd: 'resources/designSystems/withCorruptedPresets',
          params: [Command.DUMP, '--config ./uxpin.config.avatar.js', '--webpack-config "./webpack.config.js"'],
        });
      } catch (error) {
        expect(error.stderr).toMatch(/Can\'t resolve \'\.\.\/NonExistingFileWithAvatarComponent\'/gm);
      }
    });

    it('prints error about unresolved import', async () => {
      // tslint:disable-next-line:no-magic-numbers
      expect.assertions(2);

      // given
      // when
      try {
        await runUXPinMergeCommand({
          cwd: 'resources/designSystems/withCorruptedPresets',
          params: [Command.DUMP, '--config ./uxpin.config.button.js', '--webpack-config "./webpack.config.js"'],
        });
      } catch (error) {
        expect(error.stderr).toMatch(/Unknown component/gm);
        expect(error.stderr).toMatch(/src\/components\/Button\/presets\/1-corrupted-import\.jsx/gm);
      }
    });
  });
});
