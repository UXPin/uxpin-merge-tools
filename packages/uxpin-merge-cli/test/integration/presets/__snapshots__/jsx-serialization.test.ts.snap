// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The dump command run for the \`withPresets\` design system prints the JSON including serialized presets 1`] = `
Object {
  "categorizedComponents": Array [
    Object {
      "components": Array [
        Object {
          "defaultExported": true,
          "documentation": Object {
            "examples": Array [],
          },
          "info": Object {
            "dirPath": "src/components/Avatar",
            "implementation": Object {
              "framework": "reactjs",
              "lang": "typescript",
              "path": "src/components/Avatar/Avatar.tsx",
            },
            "presets": Array [
              Object {
                "path": "src/components/Avatar/presets/0-default.jsx",
              },
              Object {
                "path": "src/components/Avatar/presets/1-with func as property.jsx",
              },
            ],
          },
          "name": "Avatar",
          "presets": Array [
            Object {
              "elements": Object {
                "1": Object {
                  "name": "Avatar",
                  "props": Object {
                    "imageUrl": "https://picsum.photos/100/100?image=1027",
                    "size": "small",
                  },
                },
              },
              "name": "default",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "Avatar",
                  "props": Object {
                    "imageUrl": "https://picsum.photos/100/100?image=1027",
                  },
                },
              },
              "name": "with func as property",
              "rootId": "1",
            },
          ],
          "properties": Array [
            Object {
              "description": "",
              "isRequired": true,
              "name": "size",
              "type": Object {
                "name": "string",
                "structure": Object {},
              },
            },
            Object {
              "description": "",
              "isRequired": true,
              "name": "imageUrl",
              "type": Object {
                "name": "string",
                "structure": Object {},
              },
            },
          ],
          "wrappers": Array [],
        },
        Object {
          "defaultExported": true,
          "documentation": Object {
            "examples": Array [],
          },
          "info": Object {
            "dirPath": "src/components/ButtonWithIconAsProp",
            "implementation": Object {
              "framework": "reactjs",
              "lang": "typescript",
              "path": "src/components/ButtonWithIconAsProp/ButtonWithIconAsProp.tsx",
            },
            "presets": Array [
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/0-default.jsx",
              },
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/1-with icon.jsx",
              },
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/2-with mixed children.jsx",
              },
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/3-with missing id.jsx",
              },
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/4-with dom primitive.jsx",
              },
              Object {
                "path": "src/components/ButtonWithIconAsProp/presets/5-with renaming import.jsx",
              },
            ],
          },
          "name": "ButtonWithIconAsProp",
          "presets": Array [
            Object {
              "elements": Object {
                "1": Object {
                  "name": "ButtonWithIconAsProp",
                  "props": Object {
                    "children": "Click me",
                    "isPrimary": true,
                  },
                },
              },
              "name": "default",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "ButtonWithIconAsProp",
                  "props": Object {
                    "children": "Click me",
                    "iconStart": Object {
                      "uxpinPresetElementId": "2",
                    },
                    "isPrimary": true,
                  },
                },
                "2": Object {
                  "name": "IconStar",
                  "props": Object {
                    "stroke": "gray",
                    "strokeWidth": 2,
                  },
                },
              },
              "name": "with icon",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "ButtonWithIconAsProp",
                  "props": Object {
                    "children": Array [
                      "Add to",
                      Object {
                        "uxpinPresetElementId": "2",
                      },
                      "favourites",
                    ],
                    "isPrimary": true,
                  },
                },
                "2": Object {
                  "name": "IconStar",
                  "props": Object {
                    "stroke": "gray",
                    "strokeWidth": 2,
                  },
                },
              },
              "name": "with mixed children",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "ButtonWithIconAsProp",
                  "props": Object {
                    "children": "Add to favourites",
                    "isPrimary": true,
                  },
                },
              },
              "name": "with renaming import",
              "rootId": "1",
            },
          ],
          "properties": Array [
            Object {
              "description": "",
              "isRequired": true,
              "name": "children",
              "type": Object {
                "name": "node",
                "structure": Object {},
              },
            },
            Object {
              "description": "",
              "isRequired": false,
              "name": "iconStart",
              "type": Object {
                "name": "node",
                "structure": Object {},
              },
            },
            Object {
              "description": "",
              "isRequired": false,
              "name": "isPrimary",
              "type": Object {
                "name": "boolean",
                "structure": Object {},
              },
            },
          ],
          "wrappers": Array [],
        },
        Object {
          "defaultExported": true,
          "documentation": Object {
            "examples": Array [],
          },
          "info": Object {
            "dirPath": "src/components/MenuWithData",
            "implementation": Object {
              "framework": "reactjs",
              "lang": "typescript",
              "path": "src/components/MenuWithData/MenuWithData.tsx",
            },
            "presets": Array [
              Object {
                "path": "src/components/MenuWithData/presets/0-default.jsx",
              },
              Object {
                "path": "src/components/MenuWithData/presets/1-external data.jsx",
              },
              Object {
                "path": "src/components/MenuWithData/presets/2-with icon in data object.jsx",
              },
            ],
          },
          "name": "MenuWithData",
          "presets": Array [
            Object {
              "elements": Object {
                "1": Object {
                  "name": "MenuWithData",
                  "props": Object {
                    "data": Array [
                      Object {
                        "subtitle": "this is item 1",
                        "title": "item 1",
                      },
                      Object {
                        "subtitle": "this is item 2",
                        "title": "item 2",
                      },
                      Object {
                        "subtitle": "this is item 3",
                        "title": "item 3",
                      },
                    ],
                  },
                },
              },
              "name": "default",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "MenuWithData",
                  "props": Object {
                    "data": Array [
                      Object {
                        "subtitle": "this is another item 1",
                        "title": "item 1",
                      },
                      Object {
                        "subtitle": "this is another item 2",
                        "title": "item 2",
                      },
                    ],
                  },
                },
              },
              "name": "external data",
              "rootId": "1",
            },
            Object {
              "elements": Object {
                "1": Object {
                  "name": "MenuWithData",
                  "props": Object {
                    "data": Array [
                      Object {
                        "subtitle": "this is item 1",
                        "title": "item 1",
                      },
                      Object {
                        "subtitle": "this is item 2",
                        "title": "item 2",
                      },
                      Object {
                        "icon": Array [],
                        "subtitle": "this is item 3",
                        "title": "item 3",
                      },
                    ],
                  },
                },
              },
              "name": "with icon in data object",
              "rootId": "1",
            },
          ],
          "properties": Array [
            Object {
              "description": "",
              "isRequired": true,
              "name": "data",
              "type": Object {
                "name": "array",
                "structure": Object {},
              },
            },
          ],
          "wrappers": Array [],
        },
      ],
      "name": "General",
    },
  ],
  "name": "Example Design System",
  "vcs": ObjectContaining {
    "branchName": Any<String>,
    "commitHash": StringMatching /\\[a-z0-9\\]\\+/,
  },
}
`;
